ssl/handshake_server.cc

```diff
@@ -810,12 +812,18 @@ static enum ssl_hs_wait_t do_select_certificate(SSL_HANDSHAKE *hs) {
     return ssl_hs_error;
   }
 
+  if (!ssl_parse_client_cipher_list(&client_hello, &ssl->client_cipher_suites)) {
+    OPENSSL_PUT_ERROR(SSL, SSL_R_ERROR_IN_RECEIVED_CIPHER_LIST);
+    ssl_send_alert(ssl, SSL3_AL_FATAL, SSL_AD_HANDSHAKE_FAILURE);
+    return ssl_hs_error;
+  }
```

Comment:
Is this changing the error we return? Previously if something failed in choose_cipher or ssl_parse_client_cipher_list (which was called by choose_cipher) the below error block would be hit which sends SSL_R_NO_SHARED_CIPHER. Was this intentional, does OpenSSL return this error for this case?

Reply:
It was intentional, as I'd thought this error was more precise for describing a parse error. OpenSSL seems to [prefer](https://github.com/search?q=repo%3Aopenssl%2Fopenssl+SSL_R_ERROR_IN_RECEIVED_CIPHER_LIST&type=code) `SSL_R_ERROR_IN_RECEIVED_CIPHER_LIST` for parsing issues and [prefer](https://github.com/search?q=repo%3Aopenssl%2Fopenssl+SSL_R_NO_SHARED_CIPHER&type=code) `SSL_R_NO_SHARED_CIPHER` for not being able to find common cipher value.

Reconsidering this, though, I see the benefit of preserving backwards compatibility with local errors we return. I'll expand the places we push `SSL_R_ERROR_IN_RECEIVED_CIPHER_LIST` onto the error queue in `ssl_parse_client_cipher_list` and also push `SSL_R_NO_SHARED_CIPHER` in this failure case.